CMakeLists.txt

```
@@ -26,82 +24,162 @@ if(HIOP_USE_MPI)
 endif(HIOP_USE_MPI)
 
+# The binary dir is already a global include directory
+configure_file(
+  "${CMAKE_SOURCE_DIR}/src/Interface/hiop_defs.hpp.in"
+  "${CMAKE_BINARY_DIR}/hiop_defs.hpp")
+
+add_library(hiop_magma INTERFACE)
+add_library(hiop_math INTERFACE)
+add_library(hiop_cuda INTERFACE)
+add_library(hiop_options INTERFACE)
+add_library(hiop_warnings INTERFACE)
+
+target_compile_features(hiop_options
+    INTERFACE
+        cxx_std_11
+        cxx_alignas
+        cxx_alignof
+        cxx_attributes
+        cxx_auto_type
+        cxx_constexpr
+        cxx_defaulted_functions
+        cxx_deleted_functions
+        cxx_final
+        cxx_lambdas
+        cxx_noexcept
+        cxx_override
+        cxx_range_for
+        cxx_rvalue_references
+        cxx_static_assert
+        cxx_strong_enums
+        cxx_trailing_return_types
+        cxx_unicode_literals
+        cxx_user_literals
+        cxx_variadic_macros
+)
+
+target_compile_options(hiop_warnings
+    INTERFACE
+        -Wall
+        -Wextra
+        -Wshadow            # warn the user if a variable
+                            # declaration shadows one from a
+                            # parent context
+        -Wnull-dereference
+        -Wdouble-promotion  # Warn on implicit conversion from
+                            # float to double
+)
+
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+    message("Using extra GNU-specific compiler warnings...")
+    target_compile_options(hiop_warnings
+        INTERFACE
+            -Wmisleading-indentation    # Warn on indented blocks
+                                        # that are not really blocks
+            -Wduplicated-cond           # Warn on if/else chain with
+                                        # duplicated conditions
+            -Wduplicated-branches       # Warn on if/else chains with
+                                        # duplicated code
+            -Wuseless-cast              # Warn when casting to the same type
+    )
+endif()
+
+target_link_libraries(hiop_math INTERFACE hiop_options hiop_warnings)
+
 if(HIOP_USE_GPU)
+  include(CheckLanguage)
+  enable_language(CUDA)
+  check_language(CUDA)
+  if(NOT DEFINED CMAKE_CUDA_STANDARD)
+    set(CMAKE_CUDA_STANDARD 11)
+    set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+  endif()
+
+  if(HIOP_NVCC_ARCH)
+      set(CMAKE_CUDA_FLAGS "-arch=${HIOP_NVCC_ARCH} --expt-extended-lambda")
+      message(STATUS "Using CUDA arch ${HIOP_NVCC_ARCH}")
+  else()
+      set(CMAKE_CUDA_FLAGS "-arch=sm_35 --expt-extended-lambda")
+      message(STATUS "Using CUDA arch sm_35")
+  endif()
+
   # for now we rely on MAGMA for GPUs computations
-  set(MAGMA_LIBRARIES "-lmagma")
+  target_link_libraries(hiop_magma INTERFACE "-lmagma")
   if(HIOP_MAGMA_DIR)
-    set(MAGMA_LIBRARIES "-L${HIOP_MAGMA_DIR}/lib ${MAGMA_LIBRARIES}")
-    set(MAGMA_INCLUDE "${HIOP_MAGMA_DIR}/include")
+    target_link_libraries(hiop_magma INTERFACE "-L${HIOP_MAGMA_DIR}/lib")
```

```suggestion
    target_link_libraries(hiop_magma INTERFACE "-L${HIOP_MAGMA_DIR}/lib")
```
I think it would be good to cache `HIOP_MAGMA_DIR`, so that the configuration can be edited with `ccmake` or CMake GUI tools.

---

Good idea, I will change this.